CONFIGURATION.md

```diff
@@ -0,0 +1,297 @@
+# Configuration Design Guide
+
+## Current Status
+
+It looks like we need to have another discussion about our configuration format and how configuration behaves. What we have at the moment, looks like this:
+
+- users create a `helix-config.yaml` which will be loaded and denormalized into `.hlx/strains.json`
+- the `helix-config.yaml` contains a number of top-level (global) definitions and a map of `strains`
+- each defined strain takes default values from:
+  - the global definitions
+  - the command line options
+  - the "default" strain
+  - application defaults
+  - (the exact precedence rules differ from property to property)
+
+Our developer needs include:
+- the ability to start developing without any configuration file at all
+- the ability to quickly generate testable strains for code or content branches that are under development
+- the ability to quickly and conveniently switch between strains in simulator and production
+- the assurance that none of the above will break a published site
+- the ability to do all of the above in a fully automated fashion on a CI system (e.g. generate and deploy a new strain for a new branch)
+
+Site administration needs also include:
+- the ability to define content mount points
+- the ability to pass through traffic to another CMS (proxy strains)
+
+Furthermore, we should establish:
+- automated validation of configuration files that is independent from loading them
+- detailed and generated documentation for all configuration properties
+
+## Proposal: We keep YAML as the configuration format
+
+While I have a good collection of scars from dealing with YAML (especially in CircleCI), I think the format is moderately more readable than JSON and almost as widely accepted, which makes it worthwhile suffering through it instead of picking something that will feel uncommon or out of the time.
+
+## Proposal: We introduce a JSON Schema for configuration files
+
+This way you can see errors in configuration files while editing them, and we have a formal way of describing and documenting the configuration format. 
+
+> Notes:
+> - The validation is done during `HelixConfig.parse()`. 
+> - Using a IDE plugin, the yaml can also be validated during input.
+
+## Proposal: Strains are the only top-level concept
+
+This means that there are no concepts that are orthogonal to strains, but that every other configuration object can be assigned to one or more strains, and that every request can be assigned to one specific strain.
+
```

Comment:
I think development should work similar to the temporary strains:

1. development maps the `localhost`-URL Space to a global URL-space, so that `https://localhost:3000/www.example.com/` is treated as equivalent (in strain resolution) to `https://www.example.com/`
2. for all strains where the `code` property matches (loosely) the current working copy, use the current working directory as code base
3. for all strains where the `content` property matches (loosely) the current working copy, use the current working directory as content base
4. for all others: ??? maybe proxy OpenWhisk/GitHub?

Reply:
> development maps the localhost-URL Space to a global URL-space, so that https://localhost:3000/www.example.com/ 

I would prefer to find another way that doesn't require link rewriting. i.e. keep the HTML output identical to production. 

maybe a `hlx up --domain www.example.com`

Reply:
Good idea.